### ISSUE TITLE
User-defined types not normalized properly when passed in as data

### ISSUE THREAD

**Issue by julianyus** [select2/select2#4632]
## Prerequisites
- [x ] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [x ] The issue still exists against the latest `master` branch of Select2
- [x ] This is not a usage question (Those should be directed to the [community](https://select2.github.io/community.html))
- [x ] I have attempted to find the simplest possible steos to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)
## Steps to reproduce the issue

Initialize select 2 like this:

```
         var datos2 = [
                new producto(1, 'aaaaaa'),
                new producto(2, 'aaaaaa2'),
                new producto(3, 'aaaaaa3')
         ];

        $("#desplegable").select2({
            data: datos2
        });
```
## Expected behavior and actual behavior

When I follow those steps I see that the text of the options is: [object Object]

I was expecting it to be: aaaaaa, aaaaaa2 and aaaaaa3

According to what I have been able to investigate, the problem is here:

```
SelectAdapter.prototype._normalizeItem = function (item) {
    if (!$.isPlainObject(item)) { <--- returns false
      item = {
        id: item,
        text: item
      };
    }
...
if (item.id != null) {
      item.id = item.id.toString(); <--- [object Object]
    }
```

## Environment

Browsers
- [x ] Google Chrome
- [x ] Mozilla Firefox
- [x ] Internet Explorer

Operating System
- [x ] Windows
- [ ] Mac OS X
- [ ] Linux
- [ ] Mobile

Libraries
- jQuery version: 2.2.4
- Select2 version: 4.0.3
## Isolating the problem
- [ ] This bug happens [on the examples page](https://select2.github.io/examples.html)
- [x ] The bug happens consistently across all tested browsers
- [ ] This bug happens when using Select2 without other pluigns
- [ ] I can reproduce this bug in [a jsbin](https://jsbin.com/)

**Comment by alexweissman** [comment]
This appears to be a duplicate of #4438, which was closed due to lack of activity.  Can you create a jsbin that replicates this behavior?

I guess I don't know how jQuery decides when an object is a "plain object".  I found this documentation on [Plain objects](https://api.jquery.com/Types/#PlainObject), but it doesn't mention how it treats user-defined types (like the `producto` in your example).

**Comment by julianyus** [comment]
Hi,

I created a sample that recreates the problem: https://jsbin.com/riyamuyixi/edit?html,js,output

**Comment by alexweissman** [comment]
Thanks.  I've made this fork which shows that the bug only affects user-defined types: https://jsbin.com/cufawefalu/edit?html,js,console,output

**Comment by julianyus** [comment]
Hi,

What do you think about changing the call to isPlainObject by: 

`if(!(item === Object(item)))`

I have tested with this data and all cases have worked::

```
 var datos2 = [
                "Test",
                {
                    id: 4,
                    text: 'item'
                },
                new producto(1, 'aaaaaa'),
                new producto(2, 'aaaaaa2'),
                new producto(3, 'aaaaaa3')
         ];
```

**Comment by alexweissman** [comment]
This could be a good solution.  Keep in mind that Select2 also supports [grouped data](https://select2.org/data-sources/formats#grouped-data), so this should probably be tested as well.

Can you submit a pull request that makes the change, and also adds some additional tests for these cases?

**Comment by julianyus** [comment]
Hi,

I tested manually with grouped data and it worked

```
var datos3= [
    { 
      "text": "Group 1", 
      "children" : [
        {
            "id": 1,
            "text": "Option 1.1"
        },
        new producto(1, 'aaaaaa')
      ]
    },
    { 
      "text": "Group 2", 
      "children" : [
        {
            "id": 3,
            "text": "Option 2.1"
        },
        {
            "id": 4,
            "text": "Option 2.2"
        }
      ]
    }
  ];
```
Where dou you want me to put the tests? Should I made a new module?
I think that testing the _normalizeItem function from SelectAdapter with a string, a plain object and a user-defined object is enough because, after all, grouped data is a plain object with a property called "children".

**Comment by alexweissman** [comment]
Awesome!  So, automated tests are written with qunit, and go in https://github.com/select2/select2/tree/master/tests.

It seems like the test for user-defined types should go in either https://github.com/select2/select2/blob/master/tests/data/array-tests.js or https://github.com/select2/select2/blob/master/tests/data/select-tests.js (possibly both).

**Comment by alexweissman** [comment]
#5093 fixes this, and has been merged in to the 4.0.6-rc.0 pre-release.

**Comment by vzhilov** [comment]
In the last version 4.0.13 I had to comment that line out:

    SelectAdapter.prototype._normalizeItem = function (item) {

        if (!$.isPlainObject(item)) { <--- returns false
          item = {
            id: item,
            text: item
          };
        }
    ...
    if (item.id != null) {
    //      item.id = item.id.toString(); <--- [object Object]
        }`

 as my text object contans several fields like `text : {name="foo", lastname="bar"}` etc.
